mcpserver/server.go

```diff
@@ -0,0 +1,465 @@
+		case "add_user_to_channel":
+			if !s.config.DevMode {
+				s.logger.Debug("Tool call failed - dev mode required",
+					mlog.String("tool", toolName))
+				return &mcp.CallToolResult{
+					Content: []mcp.Content{
+						mcp.TextContent{
+							Type: "text",
+							Text: "Error: add_user_to_channel tool is only available in development mode",
+						},
+					},
+					IsError: true,
+				}, nil
+			}
+			result, err = devToolProvider.addUserToChannel(ctxWithUser, client, request.Params.Arguments)
+		case "create_post_as_user":
+			if !s.config.DevMode {
+				s.logger.Debug("Tool call failed - dev mode required",
+					mlog.String("tool", toolName))
+				return &mcp.CallToolResult{
+					Content: []mcp.Content{
+						mcp.TextContent{
+							Type: "text",
+							Text: "Error: create_post_as_user tool is only available in development mode",
+						},
+					},
+					IsError: true,
+				}, nil
+			}
+			result, err = devToolProvider.createPostAsUser(ctxWithUser, request.Params.Arguments)
+		default:
+			s.logger.Debug("Tool call failed - unknown tool",
+				mlog.String("tool", toolName))
+			return &mcp.CallToolResult{
+				Content: []mcp.Content{
+					mcp.TextContent{
+						Type: "text",
+						Text: "Error: unknown tool: " + toolName,
+					},
+				},
+				IsError: true,
+			}, nil
+		}
+
+		if err != nil {
+			s.logger.Debug("Tool call failed - execution error",
+				mlog.String("tool", toolName),
+				mlog.Err(err))
+			return &mcp.CallToolResult{
+				Content: []mcp.Content{
+					mcp.TextContent{
+						Type: "text",
+						Text: "Error: " + err.Error(),
+					},
+				},
+				IsError: true,
+			}, nil
+		}
+
+		// Log successful tool completion
+		isError := result != nil && result.IsError
+		if isError {
+			s.logger.Debug("Tool call completed with error result",
+				mlog.String("tool", toolName))
+		} else {
+			s.logger.Debug("Tool call completed successfully",
+				mlog.String("tool", toolName))
+		}
+
+		// Result is already a *mcp.CallToolResult
+		return result, nil
+	}
+}
+
+// getAuthenticatedClient gets an authenticated client for the request
+func (s *MattermostMCPServer) getAuthenticatedClient(ctx context.Context) (*model.Client4, string, error) {
+	// For OAuth mode, token must come from request context (set by HTTP transport)
+	// For PAT mode, token can come from context or fall back to config
+	var token string
+	if ctxToken, ok := ctx.Value(TokenKey).(string); ok && ctxToken != "" {
+		token = ctxToken
+	} else if s.config.PersonalAccessToken != "" {
+		// Fall back to config token for PAT mode
+		token = s.config.PersonalAccessToken
+	}
+
+	if token == "" {
+		return nil, "", fmt.Errorf("no authentication token available - ensure token is provided via context for OAuth or config for PAT")
+	}
+
+	// Create client directly - no validation needed since Mattermost APIs will validate
+	client := model.NewAPIv4Client(s.config.ServerURL)
```

Comment:
Doing this feels like maybe the job of that `AuthenticationProvider` interface. Maybe returning an authenticated APIv4Client is what that interface should do?